File: web/lib/tracking.ts
import fs from "node:fs";
import path from "node:path";
import { after } from "next/server";
import { dbEnabled, dbInsert, dbSelect } from "@/lib/server-db";

// ═══════════════════════════════════════════════════════════════════════════
// Tracking de conversão — GRAVAÇÃO (rotas /r, /k, /p) e LEITURA (painel).
//
// Antes isto vivia em tracking/events.jsonl, um arquivo no disco. Na Vercel o
// disco é efêmero: o clique era gravado e sumia no fim da invocação. O painel
// lia o mesmo arquivo e por isso mostrava zero pra sempre, e o Supervisor de
// Vendas aprendia do mesmo arquivo e por isso nunca aprendia nada.
//
// Agora o destino é a tabela `events` do Supabase. Uma regra que vale pros dois
// lados: HÁ EXATAMENTE UM DESTINO POR VEZ — com credencial, o banco; sem
// credencial (dev local), o arquivo. Nunca os dois, senão o mesmo clique
// entraria duas vezes na conta de quem lê.
// ═══════════════════════════════════════════════════════════════════════════

const ROOT = path.resolve(process.cwd(), "..");
const EVENTS = path.join(ROOT, "tracking", "events.jsonl");
const OUTPUTS = path.join(ROOT, "outputs");

// Teto de leitura do painel. O PostgREST não faz group-by sem uma view/RPC, então
// a agregação é em JS — e uma agregação em JS precisa de um teto explícito, senão
// no dia em que a tabela crescer o painel puxa a tabela inteira pra memória.
// Os mais recentes primeiro (idx_events_occurred cobre a ordenação).
const TETO_LEITURA = 5000;

export interface Ev { ts: number; event_type: string; piece: string; product_id: string; value?: number }
export interface ProdStat { product: string; clicks: number; conversions: number; value: number; cvr: number }
export interface PieceStat { piece: string; titulo: string; produto: string; clicks: number; conversions: number }

// ─────────────────────────── GRAVAÇÃO ──────────────────────────────────────

export interface DadosClique {
  /** Chave da peça: slug do dossiê, "k:GI" (ManyChat) ou "loja:bjj3d". */
  piece: string;
  /** Categoria do catálogo (ex.: gi-competicao). NÃO é id de linha da loja. */
  produto?: string | null;
  /** De onde veio o clique: peca | manychat | loja. */
  source: string;
}

/**
 * Registra um clique. CONTRATO: best-effort e NUNCA bloqueia nem quebra o
 * redirect. Se o banco cair, o usuário chega no destino do mesmo jeito.
 *
 * Como o contrato é cumprido:
 *  • com banco: a escrita vai pro `after()` do Next, que roda DEPOIS da resposta
 *    sair. O 302 não espera o Supabase — nem por um milissegundo. (O dbInsert
 *    ainda tem timeout de 2,5s pra não segurar a função da Vercel depois disso.)
 *  • sem banco: append no arquivo, como sempre foi — o fluxo local continua de pé.
 *  • o corpo inteiro é try/catch: nem um erro de programação aqui dentro pode
 *    escapar pro handler e derrubar o redirect.
 */
export function registraClique(req: Request, d: DadosClique): void {
  try {
    const url = new URL(req.url);
    const qs = (k: string) => url.searchParams.get(k) || null;
    const referrer = req.headers.get("referer") || null;
    const ua = (req.headers.get("user-agent") || "").slice(0, 200) || null;

    if (dbEnabled()) {
      after(async () => {
        await dbInsert("events", {
          event_type: "click",
          // product_id é TEXTO SOLTO desde a migração 2026-09-02-fase4-tracking.sql:
          // a FK pra products(id) foi derrubada porque products está vazia e a
          // categoria do catálogo nunca foi um id de loja. null quando não houver
          // produto — string vazia seria um id falso poluindo o group-by do painel.
          product_id: d.produto || null,
          source: d.source,
          // piece_id fica NULL de propósito. A coluna é UUID com FK pra pieces(id)
          // e o que temos aqui é um SLUG de texto — e em /k e /p nem sequer existe
          // peça ("k:GI" é palavra-chave do ManyChat, "loja:bjj3d" é produto da
          // loja). Resolver slug→UUID custaria uma consulta extra por clique pra
          // servir só um dos três casos. O slug vai em meta->>'piece', que é a
          // chave canônica de peça em todo mundo que lê (painel e Supervisor).
          meta: { piece: d.piece },
          tracked_url: url.pathname, // só o caminho: query string não vira depósito de dado
          utm_source: qs("utm_source"),
          utm_medium: qs("utm_medium"),
          utm_campaign: qs("utm_campaign"),
          utm_content: qs("utm_content"),
          referrer,
          user_agent: ua,
        });
      });
      return;
    }

    // Sem credencial (dev local): o arquivo continua sendo o destino, no MESMO
    // formato de sempre, pra que o lib/tracking.py e o painel local sigam lendo.
    const row = {
      ts: Date.now() / 1000,
      event_type: "click",
      piece: d.piece,
      product_id: d.produto || "",
      source: d.source,
      utm_source: qs("utm_source") || "",
      referrer: referrer || "",
      ua: ua || "",
    };
    fs.mkdirSync(path.dirname(EVENTS), { recursive: true });
    fs.appendFileSync(EVENTS, JSON.stringify(row) + "\n");
  } catch {
    /* clique perdido é barato; redirect preso é caro */
  }
}

// ─────────────────────────── LEITURA (painel) ──────────────────────────────

function readEvents(): Ev[] {
  if (!fs.existsSync(EVENTS)) return [];
  const out: Ev[] = [];
  for (const l of fs.readFileSync(EVENTS, "utf-8").split("\n")) {
    const s = l.trim();
    if (!s) continue;
    try { out.push(JSON.parse(s)); } catch { /* ignore */ }
  }
  return out;
}

interface LinhaEvento {
  event_type: string;
  product_id: string | null;
  value: number | string | null;
  utm_content: string | null;
  meta: { piece?: string } | null;
}

/** Lê do banco e devolve no mesmo formato do arquivo. null = ERRO de leitura. */
async function readEventsDb(): Promise<Ev[] | null> {
  const linhas = await dbSelect<LinhaEvento>(
    "events?select=event_type,product_id,value,utm_content,meta" +
    `&order=occurred_at.desc&limit=${TETO_LEITURA}`,
  );
  // dbSelect devolve null em ERRO e [] quando não achou nada — os dois NÃO podem
  // virar a mesma coisa aqui, senão banco fora do ar viraria "sem dados ainda".
  if (linhas === null) return null;
  return linhas.map((l) => ({
    ts: 0, // o painel não usa timestamp; a ordem já vem do banco
    event_type: l.event_type,
    // utm_content como segunda opção: é onde o lib/tracking.py mandava a peça
    // antes desta fase. Nenhuma linha assim existe hoje (a tabela estava vazia),
    // mas ler as duas chaves é de graça e evita um buraco silencioso.
    piece: l.meta?.piece || l.utm_content || "",
    product_id: l.product_id || "",
    value: l.value == null ? undefined : Number(l.value),
  }));
}

function tituloDe(slug: string): string {
  try {
    const m = JSON.parse(fs.readFileSync(path.join(OUTPUTS, slug, "meta.json"), "utf-8"));
    return m.dossie || slug;
  } catch {
    // Em produção a pasta outputs/ não é deployada junto com o web/, então o
    // título degrada pro slug. É legível e não custa uma consulta a mais.
    return slug;
  }
}

function agrega(evs: Ev[]) {
  const byProd = new Map<string, { clicks: number; conversions: number; value: number }>();
  const byPiece = new Map<string, { clicks: number; conversions: number; produto: string }>();
  let clicks = 0, conversions = 0, revenue = 0;
  for (const e of evs) {
    const pk = e.product_id || "—", sk = e.piece || "—";
    const p = byProd.get(pk) ?? { clicks: 0, conversions: 0, value: 0 };
    // O produto da peça vem do PRÓPRIO evento, não do meta.json em disco: assim a
    // coluna continua preenchida em produção, onde outputs/ não existe.
    const s = byPiece.get(sk) ?? { clicks: 0, conversions: 0, produto: e.product_id || "" };
    if (e.product_id && !s.produto) s.produto = e.product_id;
    if (e.event_type === "click") { p.clicks++; s.clicks++; clicks++; }
    else if (e.event_type === "conversion") {
      p.conversions++; s.conversions++; conversions++;
      p.value += Number(e.value || 0); revenue += Number(e.value || 0);
    }
    byProd.set(pk, p); byPiece.set(sk, s);
  }
  const products: ProdStat[] = [...byProd.entries()]
    .map(([product, v]) => ({ product, ...v, cvr: v.clicks ? (v.conversions / v.clicks) * 100 : 0 }))
    .sort((a, b) => b.conversions - a.conversions || b.clicks - a.clicks);
  const pieces: PieceStat[] = [...byPiece.entries()]
    .map(([piece, v]) => ({ piece, titulo: tituloDe(piece), produto: v.produto, clicks: v.clicks, conversions: v.conversions }))
    .sort((a, b) => b.clicks - a.clicks).slice(0, 12);
  return { products, pieces, totals: { clicks, conversions, revenue, cvr: clicks ? (conversions / clicks) * 100 : 0 } };
}

/**
 * Agrega pro painel. Com credencial lê do banco; sem credencial, do arquivo.
 *
 * `erro: true` significa "não consegui LER o banco" — é diferente de "não há
 * dado ainda", e o painel precisa dizer coisas diferentes nos dois casos. Tratar
 * os dois como zero é exatamente como um painel mente durante uma queda.
 */
export async function stats() {
  if (dbEnabled()) {
    const evs = await readEventsDb();
    if (evs === null) return { ...agrega([]), erro: true, fonte: "banco" as const };
    return { ...agrega(evs), erro: false, fonte: "banco" as const };
  }
  return { ...agrega(readEvents()), erro: false, fonte: "arquivo" as const };
}

File: lib/tracking.py
"""
lib/tracking.py — Tracking de conversão (o loop que faz o Supervisor aprender).

Eventos (click/conversion) chegam de dois lugares: das rotas /r, /k e /p (o
clique) e do CLI/painel (a venda). O link de rastreio registra o clique e
redireciona pro destino.

ONDE MORA O DADO. Até a fase 4 a resposta era "tracking/events.jsonl", um
arquivo no disco — e na Vercel o disco é efêmero, então em produção o clique era
gravado e sumia. A memória do Supervisor lia esse arquivo: aprendia de 4 eventos
de teste local e de mais nada. Agora a fonte é a tabela `events` do Supabase
quando há credencial, e o arquivo quando não há (dev local segue funcionando
igual, sem Supabase).

`conversion_memory()` agrega o histórico e devolve um resumo que o build_carousel
injeta no contexto do Supervisor → ele passa a dar peso ao que converte.

CLI:
  python -m lib.tracking stats
  python -m lib.tracking convert <slug> [valor]   # registra uma venda manualmente
"""
from __future__ import annotations

import json
import os
import sys
import time
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
EVENTS = ROOT / "tracking" / "events.jsonl"
OUTPUTS = ROOT / "outputs"

# Teto de leitura. A agregação é em Python, então precisa de um limite explícito:
# sem ele, no dia em que a tabela crescer, o Supervisor puxaria a tabela inteira
# a cada run. Os mais recentes primeiro.
TETO_LEITURA = 5000


def record(event_type: str, *, piece: str = "", product_id: str = "",
           value: float | None = None, **extra) -> None:
    """Grava um evento (append-only). event_type: impression | click | conversion."""
    EVENTS.parent.mkdir(parents=True, exist_ok=True)
    row = {"ts": time.time(), "event_type": event_type, "piece": piece,
           "product_id": product_id}
    if value is not None:
        row["value"] = value
    row.update({k: v for k, v in extra.items() if v is not None})
    with EVENTS.open("a", encoding="utf-8") as f:
        f.write(json.dumps(row, ensure_ascii=False) + "\n")
    try:  # dual-write best-effort no Supabase (events) — fila assíncrona do lib/db.py
        from lib import db
        db.insert_event({"event_type": event_type, "product_id": product_id or None,
                         # A peça vai em meta->>'piece', a MESMA chave que as rotas
                         # /r, /k e /p usam (web/lib/tracking.ts). Antes ia em
                         # utm_content, que quer dizer outra coisa (o parâmetro da
                         # URL) — dois vocabulários na mesma coluna. A tabela estava
                         # vazia, então dava pra acertar isso sem migrar nada.
                         "meta": {"piece": piece} if piece else None,
                         "value": value,
                         "tracked_url": extra.get("tracked_url")})
    except Exception:  # noqa: BLE001
        pass


def read_events() -> list[dict]:
    """Eventos do ARQUIVO (dev local sem Supabase)."""
    if not EVENTS.exists():
        return []
    out = []
    for line in EVENTS.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if line:
            try:
                out.append(json.loads(line))
            except json.JSONDecodeError:
                pass
    return out


def read_events_db() -> list[dict] | None:
    """
    Eventos do BANCO, no mesmo formato do arquivo. None = falha de leitura.

    A leitura é síncrona de propósito: quem chama é o Supervisor montando o
    contexto de um run, não um caminho quente de request. A fila assíncrona do
    lib/db.py existe pra ESCRITA (fire-and-forget não serve pra quem precisa da
    resposta), por isso aqui é um GET direto e com timeout curto.

    None e [] são coisas diferentes e não podem virar a mesma: [] é "não há
    evento", None é "não consegui ler". Confundir os dois faria o Supervisor
    tratar uma queda de rede como "nada converte".
    """
    url = (os.getenv("SUPABASE_URL") or "").rstrip("/")
    key = os.getenv("SUPABASE_SERVICE_ROLE_KEY") or ""
    if not (url and key):
        return None
    q = (f"{url}/rest/v1/events?select=event_type,product_id,value,utm_content,meta"
         f"&order=occurred_at.desc&limit={TETO_LEITURA}")
    req = urllib.request.Request(q, headers={
        "apikey": key, "Authorization": f"Bearer {key}", "Accept": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=5) as f:
            linhas = json.loads(f.read().decode("utf-8"))
    except Exception:  # noqa: BLE001 — o banco nunca pode derrubar o pipeline
        return None
    out = []
    for l in linhas:
        meta = l.get("meta") or {}
        out.append({
            "event_type": l.get("event_type") or "",
            "product_id": l.get("product_id") or "",
            # utm_content como segunda opção: é onde o record() mandava a peça
            # antes da fase 4. Ler as duas chaves é de graça.
            "piece": (meta.get("piece") if isinstance(meta, dict) else None)
                     or l.get("utm_content") or "",
            "value": l.get("value"),
        })
    return out


def events_para_agregar() -> list[dict]:
    """
    A fonte da verdade do momento: banco quando há credencial, arquivo quando não.

    Sempre UMA fonte só. Somar as duas contaria o mesmo clique duas vezes — o
    record() grava no arquivo E espelha no banco.
    """
    from lib import db
    if db.enabled():
        linhas = read_events_db()
        # Falha de leitura vira lista vazia aqui, e conversion_memory() devolve ""
        # — o Supervisor roda SEM a memória, que é o comportamento de quem ainda
        # não tem dado. Cair pro arquivo seria pior: ele mostraria os 4 eventos de
        # teste local como se fossem o histórico de produção.
        return linhas if linhas is not None else []
    return read_events()


def _formato(slug: str) -> str:
    p = OUTPUTS / slug / "meta.json"
    if p.exists():
        try:
            return json.loads(p.read_text(encoding="utf-8")).get("formato", "?")
        except Exception:  # noqa: BLE001
            pass
    return "?"


def aggregate() -> dict:
    """Agrega por produto: cliques, vendas, CVR. Lê do banco (ou do arquivo)."""
    agg: dict[str, dict] = {}
    for e in events_para_agregar():
        pid = e.get("product_id") or "?"
        a = agg.setdefault(pid, {"clicks": 0, "conversions": 0, "value": 0.0})
        if e.get("event_type") == "click":
            a["clicks"] += 1
        elif e.get("event_type") == "conversion":
            a["conversions"] += 1
            a["value"] += float(e.get("value") or 0)
    return agg


def conversion_memory(top: int = 8) -> str:
    """Resumo pro Supervisor. Vazio quando ainda não há dados."""
    agg = aggregate()
    if not agg or all(v["clicks"] == 0 and v["conversions"] == 0 for v in agg.values()):
        return ""
    linhas = []
    for pid, a in sorted(agg.items(), key=lambda x: (-x[1]["conversions"], -x[1]["clicks"])):
        cvr = (a["conversions"] / a["clicks"] * 100) if a["clicks"] else 0.0
        linhas.append(f"- {pid}: {a['clicks']} cliques · {a['conversions']} vendas (CVR {cvr:.0f}%)")
    return ("MEMÓRIA DE CONVERSÃO (histórico real — dê PESO ao que converte, sem trair a relevância):\n"
            + "\n".join(linhas[:top]))


if __name__ == "__main__":
    cmd = sys.argv[1] if len(sys.argv) > 1 else "stats"
    if cmd == "convert":
        slug = sys.argv[2]
        val = float(sys.argv[3]) if len(sys.argv) > 3 else None
        pid = ""
        mp = OUTPUTS / slug / "meta.json"
        if mp.exists():
            pid = json.loads(mp.read_text(encoding="utf-8")).get("produto_id", "")
        record("conversion", piece=slug, product_id=pid, value=val)
        print(f"[tracking] venda registrada: {slug} · {pid} · valor={val}")
    else:
        from lib import db
        fonte = "banco (Supabase events)" if db.enabled() else "arquivo (tracking/events.jsonl)"
        print(f"[tracking] fonte: {fonte}")  # sem isso ninguém sabe de onde veio o número
        print("[tracking] agregado:")
        print(conversion_memory() or "(sem eventos ainda)")
